crates/starknet-devnet-core/src/starknet/forked_provider.rs

```diff
@@ -0,0 +1,81 @@
+use std::str::FromStr;
+use std::sync::{mpsc, Arc};
+
+use starknet_rs_core::types::BlockId;
+use starknet_rs_ff::FieldElement;
+use starknet_rs_providers::jsonrpc::HttpTransport;
+use starknet_rs_providers::{JsonRpcClient, Provider, ProviderError};
+use url::Url;
+
+use crate::error::{DevnetResult, Error, ForkedProviderError};
+
+type ProviderResult<T> = Result<T, ProviderError>;
+
+enum ForkedRequest {
+    GetNonceAt(FieldElement, std::sync::mpsc::Sender<ProviderResult<FieldElement>>),
+}
+
+struct ForkedProvider {
+    client: Arc<JsonRpcClient<HttpTransport>>,
+    receiver: tokio::sync::mpsc::Receiver<ForkedRequest>,
+}
+
+impl ForkedProvider {
+    fn new(fork_url: Url, rx: tokio::sync::mpsc::Receiver<ForkedRequest>) -> Self {
+        let json_rpc_client = JsonRpcClient::new(HttpTransport::new(fork_url));
+
+        Self { client: Arc::new(json_rpc_client), receiver: rx }
+    }
+
+    async fn run(mut self) {
+        while let Some(msg) = self.receiver.recv().await {
+            tokio::task::spawn(Self::handle_message(msg, self.client.clone()));
+        }
+    }
+
+    async fn handle_message(msg: ForkedRequest, client: Arc<JsonRpcClient<HttpTransport>>) {
+        match msg {
+            ForkedRequest::GetNonceAt(address, sender) => {
+                let nonce = client
+                    .get_nonce(BlockId::Tag(starknet_rs_core::types::BlockTag::Latest), address)
+                    .await;
+
+                sender.send(nonce).expect("unable to send result from get_nonce");
```

Comment:
What's with this `expect`?

Reply:
This panics if the receiver end is disconnected. This means that the receiving channel crated in `ForkedOrigin::get_nonce` is dropped or went out of scope, therefore there is a bug. The other way is to ignore the result of sender.send. Error propagation is not possible, because there is no error to be returned from t `handle_message`

Reply:
Ok, I don't think we should ignore. I'm ok with the panic if it indicates a bug, just thought that maybe you forgot to change it.